src/inspector_profiler.cc

```diff
@@ -503,6 +507,215 @@ static void StopCoverage(const FunctionCallbackInfo<Value>& args) {
+    Environment* env) {
+  Local<FunctionTemplate> tmpl = env->profile_watchdog_ctor_template();
+  if (tmpl.IsEmpty()) {
+    Isolate* isolate = env->isolate();
+    tmpl = NewFunctionTemplate(isolate, nullptr);
+    tmpl->SetClassName(
+        FIXED_ONE_BYTE_STRING(env->isolate(), "ProfileWatchdog"));
+    SetProtoMethod(isolate, tmpl, "start", ProfileWatchdog::Start);
+    SetProtoMethod(isolate, tmpl, "stop", ProfileWatchdog::Stop);
+    env->set_profile_watchdog_ctor_template(tmpl);
+  }
+  return tmpl;
+}
+
+ProfileWatchdog::ProfileWatchdog(Environment* env, Local<Object> object)
+    : BaseObject(env, object) {
+  state_ = ProfileWatchdogState::kInitialized;
+  int rc;
+
+  rc = uv_loop_init(&loop_);
+  CHECK_EQ(0, rc);
+
+  rc = uv_async_init(&loop_, &async_, [](uv_async_t* task_async) {
+    ProfileWatchdog* w = ContainerOf(&ProfileWatchdog::async_, task_async);
+    w->HandleTasks();
+  });
+
+  CHECK_EQ(0, rc);
+
+  rc = uv_timer_init(&loop_, &timer_);
+  CHECK_EQ(0, rc);
+}
+
+ProfileWatchdog::~ProfileWatchdog() {
+  state_ = ProfileWatchdogState::kClosed;
+  AddTask([self = std::move(this)]() {
+    uv_close(reinterpret_cast<uv_handle_t*>(&self->timer_), nullptr);
+    uv_close(reinterpret_cast<uv_handle_t*>(&self->async_), nullptr);
+    uv_stop(&self->loop_);
+  });
+  uv_thread_join(&thread_);
+  uv_run(&loop_, UV_RUN_DEFAULT);
+  CheckedUvLoopClose(&loop_);
+}
+
+MemoryProfileWatchdog::MemoryProfileWatchdog(Environment* env,
+                                             v8::Local<v8::Object> object,
+                                             v8::Local<v8::Object> options)
+    : ProfileWatchdog(env, object) {
+  Isolate* isolate = env->isolate();
+  Local<Context> context = env->context();
+
+  Local<String> interval_key = FIXED_ONE_BYTE_STRING(isolate, "interval");
+  Local<String> max_rss_key = FIXED_ONE_BYTE_STRING(isolate, "maxRss");
+  Local<String> max_used_heap_size_key =
+      FIXED_ONE_BYTE_STRING(isolate, "maxUsedHeapSize");
+  Local<String> filename_key = FIXED_ONE_BYTE_STRING(isolate, "filename");
+
+  Local<Value> interval = options->Get(context, interval_key).ToLocalChecked();
+  interval_ = static_cast<uint64_t>(interval.As<Number>()->Value());
+
+  if (options->Has(context, max_rss_key).FromMaybe(false)) {
+    Local<Value> max_rss = options->Get(context, max_rss_key).ToLocalChecked();
+    max_rss_ = static_cast<size_t>(max_rss.As<Number>()->Value());
+  }
+
+  if (options->Has(context, max_used_heap_size_key).FromMaybe(false)) {
+    Local<Value> max_used_heap_size =
+        options->Get(context, max_used_heap_size_key).ToLocalChecked();
+    max_used_heap_size_ =
+        static_cast<size_t>(max_used_heap_size.As<Number>()->Value());
+  }
+
+  Local<Value> filename = options->Get(context, filename_key).ToLocalChecked();
+  node::Utf8Value filename_tmp(isolate, filename);
+  filename_ = std::string(*filename_tmp);
+}
+
+bool MemoryProfileWatchdog::TimeoutHandler() {
```

Comment:
Rather than this sub-classing approach, why not make it take a callback and leave it up to the user what their criteria is for stopping and do anything else they may want to do if that condition is reached? We already have basically everything in this function exposed to JavaScript already, so if we just ran a JS function and let it do its thing it'd open this up to much more easily expanding to supporting other conditions and behaviours.

Reply:
I think we can not call into JS in `RequestInterrupt` ?  the V8 doc is as follows.

>Registered |callback| must not reenter interrupted Isolate.

Reply:
Ah right, forgot about that. Could be a native callback instead of a sub-class, but not sure that's much different than what you have now. 🤷🏻